### PR TITLE
feat: add a csi-snapshotter as a sidecar into topolvm_controller

### PR DESCRIPTION
**Pull request by iamniting**
ref: https://kubernetes-csi.github.io/docs/external-snapshotter.html

Signed-off-by: Nitin Goyal <nigoyal@redhat.com>

### CODE REVIEW

File: controllers/defaults.go [hunk 1]
```diff
@@ -28,6 +28,7 @@ var (
 		"CSI_PROVISIONER_IMAGE":   "k8s.gcr.io/sig-storage/csi-provisioner:v3.0.0",
 		"CSI_LIVENESSPROBE_IMAGE": "k8s.gcr.io/sig-storage/livenessprobe:v2.5.0",
 		"CSI_RESIZER_IMAGE":       "k8s.gcr.io/sig-storage/csi-resizer:v1.3.0",
+		"CSI_SNAPSHOTER_IMAGE":    "k8s.gcr.io/sig-storage/snapshot-controller:v4.2.1",
```
Comment (Madhu-1) [inline controllers/defaults.go:31]: This is wrong, you should not use snapshot-controller you should use external-snapshotter https://kubernetes-csi.github.io/docs/external-snapshotter.html

File: controllers/topolvm_controller.go [hunk 2]
```diff
@@ -322,6 +323,22 @@ func getCsiResizerContainer() *corev1.Container {
...
 }
 
+func getCsiSnapshoterContainer() *corev1.Container {
+
+	args := []string{
+		"--v=5",
+		"--leader-election=false",
+	}
+
+	csiResizer := &corev1.Container{
```
Comment (Madhu-1) [inline controllers/topolvm_controller.go:333]: why resizer for the controller?

**Review by Madhu-1 — CHANGES_REQUESTED** [review]

File: controllers/defaults.go (hunk 1 above)
Comment (iamniting) [reply controllers/defaults.go]: Thanks for pointing it out changed it.

File: controllers/topolvm_controller.go (hunk 2 above)
Comment (iamniting) [reply controllers/topolvm_controller.go]: renamed it, copy-paste mistake.

File: controllers/topolvm_controller.go [hunk 3]
```diff
@@ -322,6 +323,28 @@ func getCsiResizerContainer() *corev1.Container {
 	return csiResizer
 }
 
+func getCsiSnapshoterContainer() *corev1.Container {
+
+	args := []string{
+		fmt.Sprintf("--csi-address=%s", DefaultCSISocket),
+		"--connection-timeout=15s",
```
Comment (Madhu-1) [inline controllers/topolvm_controller.go:330]: We dont have `connection-timeout` flag removed it from here and use the defaults?

File: controllers/topolvm_controller.go [hunk 4]
```diff
@@ -322,6 +323,28 @@ func getCsiResizerContainer() *corev1.Container {
...
+		"--leader-election=false",
+	}
+
+	volumeMounts := []corev1.VolumeMount{
+		{Name: "socket-dir", MountPath: filepath.Dir(DefaultCSISocket)},
+	}
+
+	csiSnapshoter := &corev1.Container{
```
Comment (Madhu-1) [inline controllers/topolvm_controller.go:338]: Do we dont have resource limits for this one?

File: controllers/topolvm_controller.go [hunk 5]
```diff
@@ -322,6 +323,28 @@ func getCsiResizerContainer() *corev1.Container {
 	return csiResizer
 }
 
+func getCsiSnapshoterContainer() *corev1.Container {
+
+	args := []string{
+		fmt.Sprintf("--csi-address=%s", DefaultCSISocket),
+		"--connection-timeout=15s",
+		"--leader-election=false",
```
Comment (Madhu-1) [inline controllers/topolvm_controller.go:331]: this is false by default we dont need it again here.

**Review by Madhu-1 — CHANGES_REQUESTED** [review]

File: controllers/defaults.go [hunk 6]
```diff
@@ -41,6 +42,7 @@ var (
 	CsiProvisionerImage   = GetEnvOrDefault("CSI_PROVISIONER_IMAGE")
 	CsiLivenessProbeImage = GetEnvOrDefault("CSI_LIVENESSPROBE_IMAGE")
 	CsiResizerImage       = GetEnvOrDefault("CSI_RESIZER_IMAGE")
+	CsiSnapshoterImage    = GetEnvOrDefault("CSI_SNAPSHOTER_IMAGE")
```
Comment (nbalacha) [inline controllers/defaults.go:45]:
```suggestion
	CsiSnapshotterImage    = GetEnvOrDefault("CSI_SNAPSHOTTER_IMAGE")
```

File: controllers/defaults.go [hunk 7]
```diff
@@ -57,6 +59,7 @@ var (
 	CsiRegistrarContainerName       = "csi-registrar"
 	CsiResizerContainerName         = "csi-resizer"
 	CsiProvisionerContainerName     = "csi-provisioner"
+	CsiSnapshoterContainerName      = "csi-snapshoter"
```
Comment (nbalacha) [inline controllers/defaults.go:62]:
```suggestion
	CsiSnapshotterContainerName      = "csi-snapshotter"
```

File: controllers/topolvm_controller.go [hunk 8]
```diff
@@ -124,6 +124,7 @@ func getControllerDeployment(lvmCluster *lvmv1alpha1.LVMCluster, namespace strin
 		*getCsiProvisionerContainer(),
 		*getCsiResizerContainer(),
 		*getLivenessProbeContainer(),
+		*getCsiSnapshoterContainer(),
```
Comment (nbalacha) [inline controllers/topolvm_controller.go:127]:
```suggestion
		*getCsiSnapshotterContainer(),
```

File: controllers/topolvm_controller.go [hunk 9]
```diff
@@ -322,6 +323,28 @@ func getCsiResizerContainer() *corev1.Container {
 	return csiResizer
 }
 
+func getCsiSnapshoterContainer() *corev1.Container {
```
Comment (nbalacha) [inline controllers/topolvm_controller.go:326]:
```suggestion
func getCsiSnapshotterContainer() *corev1.Container {
```

File: controllers/topolvm_controller.go [hunk 10]
```diff
@@ -322,6 +323,28 @@ func getCsiResizerContainer() *corev1.Container {
...
+		"--connection-timeout=15s",
+		"--leader-election=false",
+	}
+
+	volumeMounts := []corev1.VolumeMount{
+		{Name: "socket-dir", MountPath: filepath.Dir(DefaultCSISocket)},
+	}
+
+	csiSnapshoter := &corev1.Container{
+		Name:         CsiSnapshoterContainerName,
```
Comment (nbalacha) [inline controllers/topolvm_controller.go:339]:
```suggestion
		Name:         CsiSnapshotterContainerName,
```

File: controllers/topolvm_controller.go [hunk 11]
```diff
@@ -322,6 +323,28 @@ func getCsiResizerContainer() *corev1.Container {
...
+		Image:        CsiSnapshoterImage,
```
Comment (nbalacha) [inline controllers/topolvm_controller.go:340]:
```suggestion
		Image:        CsiSnapshotterImage,
```

File: controllers/topolvm_controller.go [hunk 12]
```diff
@@ -322,6 +323,28 @@ func getCsiResizerContainer() *corev1.Container {
...
+		Args:         args,
+		VolumeMounts: volumeMounts,
+	}
+
+	return csiSnapshoter
```
Comment (nbalacha) [inline controllers/topolvm_controller.go:345]:
```suggestion
	return csiSnapshotter
```

**Review by nbalacha — CHANGES_REQUESTED** [review]

File: controllers/topolvm_controller.go (hunk 3 above)
Comment (iamniting) [reply controllers/topolvm_controller.go]: removed it.

File: controllers/topolvm_controller.go (hunk 4 above)
Comment (iamniting) [reply controllers/topolvm_controller.go]: @nbalacha @Yuggupta27 ?

File: controllers/topolvm_controller.go (hunk 5 above)
Comment (iamniting) [reply controllers/topolvm_controller.go]: removed.

File: controllers/topolvm_controller.go (hunk 4 above)
Comment (iamniting) [reply controllers/topolvm_controller.go]: I added very minimal ones, I will change them if required.

**Review by Madhu-1 — COMMENTED** [review]
LGTM, needs to adjust the commit title and PR title.

**Comment by iamniting** [comment]
> LGTM, needs to adjust the commit title and PR title.

done

**Review by Madhu-1 — COMMENTED** [review]
forgot to mention this one.
RBAC needs to be adjusted here https://github.com/red-hat-storage/lvm-operator/blob/main/config/rbac/topolvm_controller_role.yaml#L100-L114 same as https://github.com/kubernetes-csi/external-snapshotter/blob/master/deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml#L23-L41

**Comment by iamniting** [comment]
> Please update config/default/manager_custom_env.yaml.in and the update-mgr-env Makefile target with the new ENV variable.

done

**Comment by iamniting** [comment]
> forgot to mention this one. RBAC needs to be adjusted here https://github.com/red-hat-storage/lvm-operator/blob/main/config/rbac/topolvm_controller_role.yaml#L100-L114 same as https://github.com/kubernetes-csi/external-snapshotter/blob/master/deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml#L23-L41

added

**Review by leelavg — APPROVED** [review]

File: config/rbac/topolvm_controller_role.yaml [hunk 13]
```diff
@@ -269,3 +269,44 @@ rules:
     - update
     - create
 ---
+kind: Role
```
Comment (Madhu-1) [inline config/rbac/topolvm_controller_role.yaml:272]: this should be ClusterRole not Role

File: config/rbac/topolvm_controller_role_bindings.yaml [hunk 14]
```diff
@@ -71,3 +71,15 @@ roleRef:
   name: topolvm-csi-resizer
   apiGroup: rbac.authorization.k8s.io
 ---
+kind: RoleBinding
```
Comment (Madhu-1) [inline config/rbac/topolvm_controller_role_bindings.yaml:74]: clusterRole binding

**Review by Madhu-1 — CHANGES_REQUESTED** [review]

File: config/rbac/topolvm_controller_role.yaml (hunk 13 above)
Comment (iamniting) [reply config/rbac/topolvm_controller_role.yaml]: done

File: config/rbac/topolvm_controller_role_bindings.yaml (hunk 14 above)
Comment (iamniting) [reply config/rbac/topolvm_controller_role_bindings.yaml]: done

**Review by Madhu-1 — APPROVED** [review]

**Review by nbalacha — APPROVED** [review]

**Comment by sp98** [comment]
@iamniting  Need to run `make bundle` as well.

**Comment by iamniting** [comment]
> @iamniting Need to run `make bundle` as well.

done

**Review by sp98 — APPROVED** [review]

**Pull request merged**